[FIX] runbot_travis2docker: Adding new function _ssh_keyscan to run the ssh-keyscan over the weblate-ssh server

JesusZapata opened the pull request:
When is it used the command `git fetch` over the ssh the command need a console to answer the follow question

`Are you sure you want to continue connecting (yes/no)? `

With the command `ssh-keyscan` get the key a Weblate ssh server and skip the previous question

Commit:
[FIX] runbot_travis2docker: Adding new function _ssh_keyscan to run the
ssh-keyscan over the weblate-ssh server

JesusZapata commented:
@moylop260 @ruiztulio 
This change avoid the follow question
`Are you sure you want to continue connecting (yes/no)?`

![image](https://user-images.githubusercontent.com/1387970/30381738-47464824-986b-11e7-9d23-6ef1c2ff9423.png)

ruiztulio commented:
@JesusZapata This will add the public key every time it scans it (in the runbot can be a lot)

JesusZapata commented on runbot_travis2docker/models/runbot_branch.py:
```diff
@@ -4,13 +4,33 @@
+                                        stderr=subprocess.PIPE
+                                        ).communicate()[0].split('\n')
+            for key in new_keys:
+                if [k for k in keys if k == key]:
```
@ruiztulio 
This `if` avoid that each called of the function `_ssh_keyscan` adding the same keys to the `.ssh/known_hosts` file

moylop260 commented on runbot_travis2docker/models/runbot_branch.py:
```diff
@@ -4,13 +4,33 @@
 from openerp import fields, models, api, tools
 
 
+def _ssh_keyscan(ssh):
```
Use a global flag to know if this key was created previously or LRU_CACHE

JesusZapata replied:
@moylop260 
This is better idea!

moylop260 commented on runbot_travis2docker/models/runbot_branch.py:
```diff
@@ -4,13 +4,33 @@
+                      (port_host[0], port_host[1]))
+        cmd = cmd % (port, host)
+        with open(os.path.expanduser('~/.ssh/known_hosts'), 'a+') as hosts:
+            keys = [k.replace('\n', '') for k in hosts.readlines()]
```
Use `keys = [line.strip('\n') for line in hosts]`

JesusZapata replied:
Done

moylop260 commented on runbot_travis2docker/models/runbot_branch.py:
```diff
@@ -4,13 +4,33 @@
 
 
+def _ssh_keyscan(ssh):
+    cmd = "ssh-keyscan -p %s %s"
```
Use a list instead

JesusZapata replied:
Done

moylop260 commented on runbot_travis2docker/models/runbot_branch.py:
```diff
@@ -4,13 +4,33 @@
 from openerp import fields, models, api, tools
 
 
+def _ssh_keyscan(ssh):
```
Add a docstring for what is required add a key

JesusZapata replied:
Done

moylop260 commented on runbot_travis2docker/models/runbot_branch.py:
```diff
@@ -4,13 +4,33 @@
+            keys = [k.replace('\n', '') for k in hosts.readlines()]
+            new_keys = subprocess.Popen(cmd.split(), stdout=subprocess.PIPE,
+                                        stderr=subprocess.PIPE
+                                        ).communicate()[0].split('\n')
```
How are you avoid IndexError?

JesusZapata replied:
@moylop260 
I changed the way to get the new_keys now is more secured

Commit:
[REF] runbot_travis2docker: Many fix and adding the annotation lru_cache
in order to adding cache to the function

JesusZapata commented:
@moylop260 
I made various fix referent to your comment.
Your comment is welcome

moylop260 commented on runbot_travis2docker/models/runbot_branch.py:
```diff
@@ -4,11 +4,40 @@
+        'Are you sure you want to continue connecting (yes/no)?'"""
+    cmd = ['ssh-keyscan', '-p']
+    match = re.search(r'@(?P<port_host>[^/]+)', ssh)
+    if match:
```
return early

moylop260 commented on runbot_travis2docker/models/runbot_branch.py:
```diff
@@ -4,11 +4,40 @@
+                if [line for line in hosts if (line.strip('\n') ==
+                                               key.strip('\n'))]:
+                    continue
+                hosts.write(key + '\n')
```
return True for the cache

moylop260 commented on runbot_travis2docker/models/runbot_branch.py:
```diff
@@ -4,11 +4,40 @@
+try:
+    from functools import lru_cache
+except ImportError:
+    from backports.functools_lru_cache import lru_cache
```
Add this package from manifest file

moylop260 replied:
Or use the lru cache of odoo in order to avoid install other library

JesusZapata replied:
@moylop260 
The annotation "@tools.ormcache" only work over models

I try adding the annotation but the annotation fail

![image](https://user-images.githubusercontent.com/1387970/30402883-37c23042-98ad-11e7-9a90-f6d7e50689d1.png)

Nevertheless, the code I made avoid attaching the same key several times

What do you think about delete the code and the dependency "lru_cache" in favor of using my logic?

moylop260 replied:
We have the model runbot.branch and you could use that model to add this method.

> Nevertheless, the code I made avoid attaching the same key several times

The real issue here is read many times the same file for each project * branch * cron

JesusZapata replied:
@moylop260 
Good point, I refactor the code to put the method "_ssh_keyscan" inside the class "runbot.branch"

Commit:
[REF] runbot_travis2docker: Adding new dependency
backports.functools_lru_cache and return True in _ssh_keyscan to store
the cache

Commit:
[REF] runbot_travis2docker: Delete backports.functools_lru_cache in
favor to the cache of odoo

moylop260 commented on runbot_travis2docker/models/runbot_branch.py:
```diff
@@ -33,6 +34,30 @@ def _compute_name_weblate(self):
+            'Are you sure you want to continue connecting (yes/no)?'"""
+        cmd = ['ssh-keyscan', '-p']
+        match = re.search(r'@(?P<port_host>[^/]+)', ssh)
+        if match:
```
if not match: return False

JesusZapata replied:
Done

moylop260 commented on runbot_travis2docker/models/runbot_branch.py:
```diff
@@ -33,6 +34,30 @@ def _compute_name_weblate(self):
+        cmd = ['ssh-keyscan', '-p']
+        match = re.search(r'@(?P<port_host>[^/]+)', ssh)
+        if match:
+            port_host = match.groupdict()['port_host'].split(':')
```
What about use regex to get `host:port`? (instead of split(':')

JesusZapata replied:
@moylop260 
I don't known how to write a regular expression to avoid the function split.
I need more knowledge in this matter, sorry :disappointed: 
Can you help me with this regular expression ?

Commit:
[REF] runbot_travis2docker: Return early in the if statement

Commit:
[REF] runbot_travis2docker: Better use one regular expression to avoid
the split function and many if statements

JesusZapata commented on runbot_travis2docker/models/runbot_branch.py:
```diff
@@ -41,13 +41,10 @@ def _ssh_keyscan(self, ssh):
             'Are you sure you want to continue connecting (yes/no)?'"""
         cmd = ['ssh-keyscan', '-p']
-        match = re.search(r'@(?P<port_host>[^/]+)', ssh)
+        match = re.search(r'@([^/]+(?=:|/]))?(?::)?([^/]+)?/', ssh)
```
@moylop260 
Good news! 
I found a regular expression to avoid a split function and another if statements

Commit:
[REF] runbot_travis2docker: Adding name to the group in favor to get
more information

JesusZapata commented on runbot_travis2docker/models/runbot_branch.py:
```diff
@@ -41,7 +41,8 @@ def _ssh_keyscan(self, ssh):
             'Are you sure you want to continue connecting (yes/no)?'"""
         cmd = ['ssh-keyscan', '-p']
-        match = re.search(r'@([^/]+(?=:|/]))?(?::)?([^/]+)?/', ssh)
+        match = re.search(r'@(?P<host>[^/]+(?=:|/]))?(?::)?(?P<port>[^/]+)?/',
```
@moylop260 
I try to use the expression regular existing within the function "get_repo_data" but this not work as should be

![image](https://user-images.githubusercontent.com/1387970/30464124-a7961d2a-999d-11e7-8824-d38983461a5d.png)

I add group name to the regular expression existing in this code

![image](https://user-images.githubusercontent.com/1387970/30464001-fd1d232a-999c-11e7-859a-6016cbefecb4.png)

Only one trick the function "sorted" is required to get first the port and the second one the host

moylop260 commented on runbot_travis2docker/models/runbot_branch.py:
```diff
@@ -33,6 +34,29 @@ def _compute_name_weblate(self):
+                          ssh)
+        if not match:
+            return False
+        cmd.extend(sorted(match.groups('22')))
```
You are using named sections so you can get the value here with that  name

JesusZapata replied:
@moylop260 
Sorry for my short explanation before

![image](https://user-images.githubusercontent.com/1387970/30492661-3dec38c4-9a0f-11e7-910f-38ecb2bc0027.png)

The expression regular made for me has a little bug when the port is not present into the url the group returned has bad position 

For that reason the function "sorted" is required for to fix my little trouble

Sorry that was the only way I found :disappointed:

moylop260 replied:
What about use `@(?P<host>[^/]+(?=:|/]))?(?::)?(?P<port>\d+)([^/]+)?/'`?
I mean, for the port your get just digits with `\d+`

JesusZapata replied:
@moylop260 

I try to use this regular expression, but when the group "port" not exists the regular expression fail

![image](https://user-images.githubusercontent.com/1387970/30497431-da93e936-9a20-11e7-9ebd-7b2dc994c4ba.png)

In this URL you can test the regular expresison 
https://regex101.com/r/GA0BJ7/1

moylop260 replied:
So, you need change your regex to support both cases.

Something like as:
https://regex101.com/r/qLelXK/3/

JesusZapata replied:
@moylop260 
That regular expression is awesome.

moylop260 commented on runbot_travis2docker/models/runbot_branch.py:
```diff
@@ -33,6 +34,29 @@ def _compute_name_weblate(self):
+        with open(os.path.expanduser('~/.ssh/known_hosts'), 'a+') as hosts:
+            new_keys = subprocess.Popen(cmd, stdout=subprocess.PIPE,
+                                        stderr=subprocess.PIPE
+                                        ).stdout.readlines()
```
Avoid use `readlines` in order to save memory ram.
You can use
`for line in stdout:`

JesusZapata replied:
Good point, to optimize the memory use

Commit:
[FIX] runbot_travis2docker: Avoid read all lines into the file and read
each line one time

Commit:
[REF] runbot_travis2docker: Change the regular expression to another one
better

JesusZapata commented:
@moylop260 
I change the regular expression to your regular expression. 
Thanks for you help.

Commit:
[REF] runbot_travis2docker: Change the way to get the port for default
to be more explicit
